### PR TITLE
java-backend: support SMTLIB translation for KVar binders

### PR DESCRIPTION
**Pull request by daejunpark**
~Support the format `[smtlib((forall|exists ((#1 <sort>)) <term>))]`. See the test for an example.~

Support SMTLIB translation for binders of sort `KVar`.

### CODE REVIEW

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java [hunk 1]
```diff
@@ -656,6 +676,13 @@ public SMTLibTerm transform(BitVector bitVector) {
 
     @Override
     public SMTLibTerm transform(UninterpretedToken uninterpretedToken) {
+        if (uninterpretedToken.sort() == Sort.KVARIABLE) {
+            if (binders.search(uninterpretedToken) != -1) {
```
Comment (denis-bogdanas) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java]: contains() must be used here

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java [hunk 2]
```diff
@@ -571,10 +587,14 @@ public SMTLibTerm transform(KItem kItem) {
         if (label.startsWith("(")) {
             // smtlib expression instead of operator
             String expression = label;
+            boolean hasBinder = recordBinders(label, kList);
```
Comment (denis-bogdanas) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java]: This whole feature is extraordinary error-prone. I'd say a warning must be added here when it is used.

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java [hunk 3]
```diff
@@ -243,6 +246,7 @@ public static CharSequence translateImplication(
     private final LinkedHashSet<Variable> variables;
     private final LinkedHashMap<Term, Variable> termAbstractionMap;
     private final LinkedHashMap<UninterpretedToken, Integer> tokenEncoding;
+    private final Stack<Term> binders;
```
Comment (denis-bogdanas) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java:247]: This variable must be renamed. Like `smtlibForallOrExistsBinder` ? Long names are better than confusing names.

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java (hunk 1 above)
Comment (daejunpark) [reply java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java]: fixed

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java (hunk 2 above)
Comment (daejunpark) [reply java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java]: added

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java (hunk 3 above)
Comment (daejunpark) [reply java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java:247]: done

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java (hunk 2 above)
Comment (daejunpark) [reply java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java]: This is no longer applicable.

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java (hunk 3 above)
Comment (daejunpark) [reply java-backend/src/main/java/org/kframework/backend/java/symbolic/KILtoSMTLib.java:247]: Now this PR becomes more general, and the name `binders` is appropriate.

**Comment by daejunpark** [comment]
Now this PR becomes more general, supporting SMTLIB translation for arbitrary KVar binders. Please re-review.

**Review by denis-bogdanas — APPROVED** [review]

**Pull request merged**